### PR TITLE
010 -- Pullfile Complement to Pushfile

### PR DESCRIPTION
**Pull request by rakuy0**

### CODE REVIEW

File: synapse/tests/test_tools_pullfile.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import synapse.tools.pushfile as s_pushfile
+import synapse.tools.pullfile as s_pullfile
+
+
+class TestPullFile(s_t_utils.SynTest):
+
+    async def test_pullfile(self):
+
+        async with self.getTestDmonCortexAxon() as dmon:  # NOQA
```
Comment (vEpiphyte) [inline synapse/tests/test_tools_pullfile.py]: There is no need for the  `# NOQA` here

File: synapse/tools/pullfile.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        except Exception as e:
+            outp.printf('Error: Hit Exception: %s' % (str(e),))
+            continue
+
+    s_glob.sync(axon.fini())
+    return 0
+
+
+def setup():
+    desc = 'Looks up a file in a cortex and then fetches it from the given axon'
```
Comment (vEpiphyte) [inline synapse/tools/pullfile.py]: This doesn't lookup files against a cortex.

File: synapse/tools/pullfile.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+'''
+Fetches file details from a cortex and then pulls the associated file from the axon given
```
Comment (vEpiphyte) [inline synapse/tools/pullfile.py]: cortex is not used here

File: synapse/tools/pullfile.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            outp.printf(f'Fetching {h} to file')
+            with open(f'{opts.output}/{h}', 'wb') as fd:
+                for b in axon.get(binascii.unhexlify(h)):
+                    fd.write(b)
+            outp.printf(f'Fetched {h} to file')
+        except Exception as e:
+            outp.printf('Error: Hit Exception: %s' % (str(e),))
+            continue
+
+    s_glob.sync(axon.fini())
```
Comment (vEpiphyte) [inline synapse/tools/pullfile.py]: You can open the proxy as a context manager and avoid having to do this explicit sync() call.

File: synapse/tools/pullfile.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    if outp is None:
+        outp = s_output.OutPut()
+
+    if opts.output is None:
+        opts.output = '.'
+
+    s_common.gendir(opts.output)
+
+    axon = s_telepath.openurl(opts.axon)
```
Comment (jnwatson) [inline synapse/tools/pullfile.py]: `with s_telepath.openurl(opts.axon) as axon:` would be slightly nicer and preclude having to fini at the end.

File: synapse/tools/pullfile.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    for a in awants:
+        outp.printf(f'{binascii.hexlify(a)} not in axon store')
+
+    exists = [h for h in opts.hashes if binascii.unhexlify(h) not in awants]
+
+    for h in exists:
+
+        try:
+            outp.printf(f'Fetching {h} to file')
+            with open(f'{opts.output}/{h}', 'wb') as fd:
```
Comment (jnwatson) [inline synapse/tools/pullfile.py]: plz use pathlib or os.path.join for filename manipulation (prefer pathlib).

**Pull request merged**